include/lock/optimistic_lock.hpp

```diff
@@ -0,0 +1,233 @@
+#include <utility>
+
+namespace dbgroup::lock
+{
+class OptimisticLock
+{
+ public:
+  /*####################################################################################
+   * Public constructors and assignment operators
+   *##################################################################################*/
+
+  constexpr OptimisticLock() = default;
+
+  OptimisticLock(const OptimisticLock&) = delete;
+  OptimisticLock(OptimisticLock&&) = delete;
+
+  auto operator=(const OptimisticLock&) -> OptimisticLock& = delete;
+  auto operator=(OptimisticLock&&) -> OptimisticLock& = delete;
+
+  /*####################################################################################
+   * Public destructors
+   *##################################################################################*/
+
+  ~OptimisticLock() = default;
+
+  /*####################################################################################
+   * Public getters
+   *##################################################################################*/
+
+  /**
+   * @brief Get the Version object
+   *
+   * @return std::pair<uint64_t, bool>
+   */
+  [[nodiscard]] auto
+  GetVersion()  //
+      -> std::pair<uint64_t, bool>
+  {
+    version_ = lock_.load(std::memory_order_relaxed);
+    if (version_ & (2UL) == 0b10) return std::make_pair(version_, false);
+    return std::make_pair(version_, true);
+  }
+
+  /**
+   * @brief Check whether the version after reading is equal to the version before reading
+   *
+   * @return true
+   * @return false
+   */
+  auto
+  CheckVersion()  //
+      -> bool
+  {
+    std::pair<uint64_t, bool> current_version_ = GetVersion();
+    if (!GetVersion().second) return false;
+    if (current_version_.first == version_) return true;
+    return false;
+  }
+
+  /**
+   * @brief Get an exclusive lock.
+   *
+   * NOTE: this function does not give up acquiring locks and continues spinlock.
+   */
+  void
+  LockX()
+  {
+    auto expected = lock_.load(std::memory_order_relaxed);
+    auto desired = expected + kXLock;
+
+    while (!lock_.compare_exchange_weak(expected, desired, std::memory_order_acquire,
+                                        std::memory_order_relaxed)) {
+      desired = expected + kXLock;
+      SPINLOCK_HINT
+    }
+  }
+
+  /**
+   * @brief Downgrade an X lock to an SIX lock.
+   *
+   * NOTE: if a thread that does not have an exclusive lock calls this function, it will
+   * corrupt an internal lock status.
+   */
+  void
+  DowngradeToSIX()
+  {
+    lock_.store(kSIXLock, std::memory_order_release);
+  }
+
+  /**
+   * @brief Release an exclusive lock.
+   *
+   */
+  void
+  UnlockX()
+  {
+    auto expected = lock_.load(std::memory_order_relaxed);
+    auto desired = expected + kXLock;
+
+    while (!lock_.compare_exchange_weak(expected, desired, std::memory_order_acquire,
+                                        std::memory_order_relaxed)) {
+      desired = expected + kXLock;
+      SPINLOCK_HINT
+    }
```

Comment:
こっちも`fetch_add`でOKです．

```suggestion
    lock_.fetch_add(kXLock, std::memory_order_release);
```